editor/src/clj/internal/node.clj

```diff
@@ -1334,11 +1334,14 @@
     (type-name (gt/node-type node))))
 
 (defn mark-in-production [node-id label evaluation-context]
-  (assert (not (contains? (:in-production evaluation-context) (gt/endpoint node-id label)))
-          (format "Cycle detected on node type %s and output %s"
-                  (node-type-name node-id evaluation-context)
-                  label))
-  (update evaluation-context :in-production conj (gt/endpoint node-id label)))
+  (if (contains? (:in-production evaluation-context) (gt/endpoint node-id label))
+    (throw (ex-info (format "Cycle detected on node type %s and output %s"
+                      (node-type-name node-id evaluation-context)
+                      label)
+                    {:cause "cycle-detected"
+                     :node-type-name (node-type-name node-id evaluation-context)
```

Comment:
I don't think we need to add the `:node-type-name` here. We can get the node type from the `node-id` when we generate the error message in `get-cycle-detected-help-message` above.